vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareContextPool.java

```diff
@@ -45,76 +43,74 @@ public VmwareContextPool() {
+        final String poolKey = composePoolKey(vCenterAddress, vCenterUserName);
+        if (Strings.isNullOrEmpty(poolKey)) {
+            return null;
         }
-    }
-
-    public VmwareContext getContext(String vCenterAddress, String vCenterUserName) {
-        String poolKey = composePoolKey(vCenterAddress, vCenterUserName);
         synchronized (this) {
-            List<VmwareContext> l = _pool.get(poolKey);
-            if (l == null)
-                return null;
-
-            if (l.size() > 0) {
-                VmwareContext context = l.remove(0);
-                context.setPoolInfo(this, poolKey);
-
-                if (s_logger.isTraceEnabled())
-                    s_logger.trace("Return a VmwareContext from the idle pool: " + poolKey + ". current pool size: " + l.size() + ", outstanding count: " +
-                        VmwareContext.getOutstandingContextCount());
+            final Queue<VmwareContext> ctxList = _pool.get(poolKey);
+            if (ctxList != null && ctxList.size() > 0) {
+                final VmwareContext context = ctxList.remove();
+                if (context != null) {
+                    context.setPoolInfo(this, poolKey);
+                }
+                if (s_logger.isTraceEnabled()) {
+                    s_logger.trace("Return a VmwareContext from the idle pool: " + poolKey + ". current pool size: " + ctxList.size() + ", outstanding count: " +
+                            VmwareContext.getOutstandingContextCount());
+                }
                 return context;
             }
-
-            // TODO, we need to control the maximum number of outstanding VmwareContext object in the future
             return null;
         }
     }
 
-    public void returnContext(VmwareContext context) {
+    public void registerContext(final VmwareContext context) {
         assert (context.getPool() == this);
         assert (context.getPoolKey() != null);
         synchronized (this) {
-            List<VmwareContext> l = _pool.get(context.getPoolKey());
-            if (l == null) {
-                l = new ArrayList<VmwareContext>();
-                _pool.put(context.getPoolKey(), l);
+            Queue<VmwareContext> ctxList = _pool.get(context.getPoolKey());
+            if (ctxList == null) {
+                ctxList = EvictingQueue.create(_maxIdleQueueLength);
+                _pool.put(context.getPoolKey(), ctxList);
             }
 
-            if (l.size() < _maxIdleQueueLength) {
-                context.clearStockObjects();
-                l.add(context);
+            if (ctxList.size() > _maxIdleQueueLength) {
+                final VmwareContext oldestContext = ctxList.remove();
+                if (oldestContext != null) {
+                    try {
+                        oldestContext.close();
+                    } catch (Throwable t) {
+                        s_logger.error("Unexpected exception caught while trying to purge oldest VmwareContext", t);
+                    }
```

Comment:
Although we are using a bounded queue now and that will prevent memory leaks, do we know why it should be throwing old elements -or- why the contexts are not cleaned up programmatically instead of forcefully restricting the size ?